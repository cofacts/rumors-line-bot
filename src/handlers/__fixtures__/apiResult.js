const selectedArticleId = {
  data: {
    GetArticle: {
      replyCount: 2,
      replyConnections: [
        {
          id: 'AVyyB61NyCdS-nWhuakC__AVygFA0RyCdS-nWhuaXY',
          reply: {
            id: 'AVygFA0RyCdS-nWhuaXY',
            versions: [
              {
                type: 'RUMOR',
                text: '查證這則謠言的過程中，約翰走鹿也發現其實網路早就有許多解答，就讓約翰走鹿來借花獻佛一下囉。多家媒體都有針對這則「吃生魚片會有海獸胃腺蟲附著胃壁」的謠言請教醫師，受訪的醫師們皆表示，這則謠言太誇大，海獸胃腺蟲沒那麼神通廣大。\n醫師們表示，台灣的衛生條件很進步，吃生魚片被寄生蟲寄生的案例很少，大部份都是生魚片不乾淨，感染沙門氏菌、大腸桿菌的案例，此外，寄生蟲也不是什麼難治之症，服用一些抗寄生蟲藥就可以殺死胃裡的這些蟲子，除非這些蟲子跑到身體其他部位那才比較難處理，但那也是要大量地吃生魚片（一週兩次以上），中獎機率才會比較高。',
              },
            ],
          },
          feedbacks: [
            {
              comment: null,
              score: 1,
            },
            {
              comment: null,
              score: -1,
            },
            {
              comment: null,
              score: 1,
            },
            {
              comment: null,
              score: 1,
            },
            {
              comment: null,
              score: 1,
            },
            {
              comment: null,
              score: 1,
            },
            {
              comment: null,
              score: 1,
            },
            {
              comment: null,
              score: -1,
            },
            {
              comment: null,
              score: 1,
            },
            {
              comment: null,
              score: 1,
            },
            {
              comment: null,
              score: 1,
            },
            {
              comment: null,
              score: 1,
            },
            {
              comment: null,
              score: 1,
            },
            {
              comment: null,
              score: 1,
            },
            {
              comment: null,
              score: 1,
            },
            {
              comment: null,
              score: 1,
            },
          ],
        },
        {
          id: 'AVyyB61NyCdS-nWhuakC__AVy6LkWIyCdS-nWhuaqu',
          reply: {
            id: 'AVy6LkWIyCdS-nWhuaqu',
            versions: [
              {
                type: 'RUMOR',
                text: '過去台灣只有一個海獸胃腺蟲鑽入胃壁的病例，而且僅有一隻蟲。而經查詢，台北馬偕醫院也沒有名為林清風的醫師。',
              },
            ],
          },
          feedbacks: [
            {
              comment: null,
              score: 1,
            },
            {
              comment: null,
              score: 1,
            },
            {
              comment: null,
              score: 1,
            },
            {
              comment: null,
              score: 1,
            },
            {
              comment: null,
              score: 1,
            },
            {
              comment: null,
              score: 1,
            },
            {
              comment: null,
              score: 1,
            },
            {
              comment: null,
              score: 1,
            },
            {
              comment: null,
              score: 1,
            },
            {
              comment: null,
              score: 1,
            },
            {
              comment: null,
              score: 1,
            },
            {
              comment: null,
              score: 1,
            },
            {
              comment: null,
              score: 1,
            },
            {
              comment: null,
              score: 1,
            },
            {
              comment: null,
              score: 1,
            },
            {
              comment: null,
              score: 1,
            },
            {
              comment: null,
              score: 1,
            },
            {
              comment: null,
              score: 1,
            },
            {
              comment: null,
              score: 1,
            },
            {
              comment: null,
              score: 1,
            },
            {
              comment: null,
              score: 1,
            },
            {
              comment: null,
              score: 1,
            },
            {
              comment: null,
              score: 1,
            },
            {
              comment: null,
              score: 1,
            },
          ],
        },
      ],
    },
  },
};

const multipleReplies = {
  data: {
    GetArticle: {
      replyCount: 4,
      replyConnections: [
        {
          id: 'AV8d2-YtyCdS-nWhuhdi__AV--O3nfyCdS-nWhujMD',
          reply: {
            id: 'AV--O3nfyCdS-nWhujMD',
            versions: [
              {
                type: 'RUMOR',
                text: '老榮民有住所，有月退俸，也不需要特別的醫療護理，因此較適合榮民之家養老，慈濟也一直關懷陪伴。護理之家有免費方案，入住資格和有無捐款沒有關係。\n慈善機構使用資源需符合規定，不能僅為少數人開特例。不是曾捐款就需要特別厚待，不是未曾捐款就可以不理。',
              },
            ],
          },
          feedbacks: [
            {
              comment: null,
              score: 1,
            },
            {
              comment: null,
              score: 1,
            },
            {
              comment: null,
              score: 1,
            },
          ],
        },
        {
          id: 'AV8d2-YtyCdS-nWhuhdi__AV--LRZYyCdS-nWhujL9',
          reply: {
            id: 'AV--LRZYyCdS-nWhujL9',
            versions: [
              {
                type: 'NOT_ARTICLE',
                text: '整個基督教徒的打手，自己不努力去牧人、募款，利用攻擊別人，三分真七分假，更利用多帳號來造勢及造風向，害我也差點相信了，各位有智慧的看官，看過之後，笑一笑即可，別太認真去相信他的話啦。\n',
              },
            ],
          },
          feedbacks: [],
        },
        {
          id: 'AV8d2-YtyCdS-nWhuhdi__AV8jkRlByCdS-nWhuhiY',
          reply: {
            id: 'AV8jkRlByCdS-nWhuhiY',
            versions: [
              {
                type: 'NOT_ARTICLE',
                text: '文章是假的 照片也是盜用我方家人的照片,嚴重影響我家人名譽,我方已對po文的人聯絡並打算提出告訴,請勿相信謠言.',
              },
            ],
          },
          feedbacks: [],
        },
        {
          id: 'AV8d2-YtyCdS-nWhuhdi__AV8fXikZyCdS-nWhuhfN',
          reply: {
            id: 'AV8fXikZyCdS-nWhuhfN',
            versions: [
              {
                type: 'OPINIONATED',
                text: '慈濟自來有褒有貶，每人有各自的看法。',
              },
            ],
          },
          feedbacks: [],
        },
      ],
    },
  },
};
export default { selectedArticleId, multipleReplies };
