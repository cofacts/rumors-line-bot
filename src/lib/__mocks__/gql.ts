const mockResultQueue: object[] = [];

function gqlMock() {
  return () => {
    const ret = mockResultQueue.shift();
    return Promise.resolve(ret);
  };
}

/**
 * Enqueues mock data what will be returned in next gql()() call.
 *
 * Once gql()() is invoked, the first data in queue will be consumed.
 *
 * @param {*} returnValue the mock value that gql()()'s returned promise will resolve to.
 * @returns nothing
 */
gqlMock.__push = function (returnValue: object) {
  mockResultQueue.push(returnValue);
};

/**
 * Empty the return data queue.
 *
 * Invoking this in afterEach() will ensure 1 failed test would not leave side effects to other
 * tests.
 */
gqlMock.__reset = function () {
  mockResultQueue.length = 0;
};

/**
 * @returns {Boolean} if the mock response has been depleted
 */
gqlMock.__finished = function () {
  return mockResultQueue.length === 0;
};

export default gqlMock;
export type MockedGql = typeof gqlMock;
