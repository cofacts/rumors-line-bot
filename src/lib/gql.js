import fetch from 'node-fetch';
import rollbar from './rollbar';
import url from 'url';

const API_URL =
  process.env.API_URL || 'https://cofacts-api.hacktabl.org/graphql';

// Usage:
//
// import gql from './util/GraphQL';
// gql`query($var: Type) { foo }`({var: 123}).then(...)
//
// gql`...`() returns a promise that resolves to immutable Map({data, errors}).
//
// We use template string here so that Atom's language-babel does syntax highlight
// for us.
//
// GraphQL Protocol: http://dev.apollodata.com/tools/graphql-server/requests.html
//
export default (query, ...substitutions) => (variables, search) => {
  const queryAndVariable = {
    query: String.raw(query, ...substitutions),
  };

  if (variables) queryAndVariable.variables = variables;

  let status;
  const URL = `${API_URL}${url.format({ query: search })}`;

  return fetch(URL, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      'x-app-secret': process.env.APP_SECRET,
    },
    credentials: 'include',
    body: JSON.stringify(queryAndVariable),
  })
    .then(r => {
      status = r.status;
      return r.json();
    })
    .then(resp => {
      if (status === 400) {
        throw new Error(
          `GraphQL Error: ${resp.errors
            .map(({ message }) => message)
            .join('\n')}`
        );
      }
      if (resp.errors) {
        // When status is 200 but have error, just print them out.
        console.error('GraphQL operation contains error:', resp.errors);
        rollbar.error(
          'GraphQL error',
          {
            body: JSON.stringify(queryAndVariable),
            url: URL,
          },
          { resp }
        );
      }
      return resp;
    });
};
