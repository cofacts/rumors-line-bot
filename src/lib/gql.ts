import fetch from 'node-fetch';
import rollbar from './rollbar';
import { format } from 'url';
import Dataloader from 'dataloader';

const API_URL = process.env.API_URL || 'https://dev-api.cofacts.tw/graphql';

type QV = {
  query: string;
  variables?: object;
};

type Resp = { data: object | null; errors: { message: string }[] | undefined };

// Maps URL to dataloader. Cleared after batched request is fired.
// Exported just for unit test.
export const loaders: Record<string, Dataloader<QV, Resp>> = {};

/**
 * Returns a dataloader instance that can send query & variable to the GraphQL endpoint specified by `url`.
 *
 * The dataloader instance is automatically created when not exist for the specified `url`, and is
 * cleared automatically when the batch request fires.
 *
 * @param url - GraphQL endpoint URL
 * @returns A dataloader instance that loads response of the given {query, variable}
 */
function getGraphQLRespLoader(url: string) {
  if (loaders[url]) return loaders[url];

  return (loaders[url] = new Dataloader<QV, Resp>(async (queryAndVariables) => {
    // Clear dataloader so that next batch will get a fresh dataloader
    delete loaders[url];

    // Implements Apollo's transport layer batching
    // https://www.apollographql.com/blog/apollo-client/performance/query-batching/#1bce
    //
    return (
      await fetch(url, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'x-app-secret': process.env.APP_SECRET ?? '',
        },
        body: JSON.stringify(queryAndVariables),
      })
    ).json();
  }));
}

// Usage:
//
// import gql from './util/GraphQL';
// gql`query($var: Type) { foo }`({var: 123}).then(...)
//
// gql`...`() returns a promise that resolves to immutable Map({data, errors}).
//
// We use template string here so that Atom's language-babel does syntax highlight
// for us.
//
export default (query: TemplateStringsArray, ...substitutions: string[]) =>
  <QueryResp extends object, Variable>(
    variables: Variable,
    search?: Record<string, string | number>
  ) => {
    const queryAndVariable: QV = {
      query: String.raw(query, ...substitutions),
    };

    if (variables) queryAndVariable.variables = variables;

    const URL = `${API_URL}${format({ query: search })}`;

    return getGraphQLRespLoader(URL)
      .load(queryAndVariable)
      .then((resp) => {
        // We cannot get status code of individual request in transport layer batching.
        // but we can guess that it's not 2xx if `data` is null or does not exist.
        // Ref: https://github.com/graphql/graphql-over-http/blob/main/spec/GraphQLOverHTTP.md#status-codes
        //
        if (!('data' in resp) || !resp.data || typeof resp.data !== 'object') {
          const errorStr =
            'errors' in resp && Array.isArray(resp.errors)
              ? resp.errors.map(({ message }) => message).join('\n')
              : 'Unknown error';

          throw new Error(`GraphQL Error: ${errorStr}`);
        }

        if ('errors' in resp && resp.errors) {
          console.error('GraphQL operation contains error:', resp.errors);
          rollbar.error(
            'GraphQL error',
            {
              body: JSON.stringify(queryAndVariable),
              url: URL,
            },
            { resp }
          );
        }

        return resp as GqlResponse<QueryResp>;
      });
  };
