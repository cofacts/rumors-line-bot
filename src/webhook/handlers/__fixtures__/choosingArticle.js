export const selectedArticleId = {
  data: {
    GetArticle: {
      text: '《緊急通知》\n台北馬偕醫院傳來訊息：\n資深醫生（林清風）傳來：「請大家以後千萬不要再吃生魚片了！」\n因為最近已經發現- 好多病人因為吃了生魚片，胃壁附著《海獸胃腺蟲》，大小隻不一定，有的病人甚至胃壁上滿滿都是無法夾出來，驅蟲藥也很難根治，罹患機率每個國家的人都一樣。\n尤其；鮭魚的含蟲量最高、最可怕！\n請傳給朋友，讓他們有所警惕!',
      replyCount: 2,
      articleReplies: [
        {
          reply: {
            id: 'AVygFA0RyCdS-nWhuaXY',
            type: 'RUMOR',
            text: '查證這則謠言的過程中，約翰走鹿也發現其實網路早就有許多解答，就讓約翰走鹿來借花獻佛一下囉。多家媒體都有針對這則「吃生魚片會有海獸胃腺蟲附著胃壁」的謠言請教醫師，受訪的醫師們皆表示，這則謠言太誇大，海獸胃腺蟲沒那麼神通廣大。\n醫師們表示，台灣的衛生條件很進步，吃生魚片被寄生蟲寄生的案例很少，大部份都是生魚片不乾淨，感染沙門氏菌、大腸桿菌的案例，此外，寄生蟲也不是什麼難治之症，服用一些抗寄生蟲藥就可以殺死胃裡的這些蟲子，除非這些蟲子跑到身體其他部位那才比較難處理，但那也是要大量地吃生魚片（一週兩次以上），中獎機率才會比較高。',
          },
          positiveFeedbackCount: 14,
          negativeFeedbackCount: 2,
        },
        {
          reply: {
            id: 'AVy6LkWIyCdS-nWhuaqu',
            type: 'RUMOR',
            text: '過去台灣只有一個海獸胃腺蟲鑽入胃壁的病例，而且僅有一隻蟲。而經查詢，台北馬偕醫院也沒有名為林清風的醫師。',
          },
          positiveFeedbackCount: 24,
          negativeFeedbackCount: 0,
        },
      ],
    },
  },
};

export const oneReply = {
  data: {
    GetArticle: {
      replyCount: 1,
      text: 'Just One Reply Just One Reply Just One Reply Just One Reply Just One Reply Just One Reply Just One Reply Just One Reply Just One Reply Just One Reply ',
      articleReplies: [
        {
          reply: {
            id: 'AV--O3nfyCdS-nWhujMD',
            type: 'RUMOR',
            text: 'The Only Reply The Only Reply The Only Reply The Only Reply The Only Reply The Only Reply The Only Reply The Only Reply The Only Reply ',
          },
          positiveFeedbackCount: 3,
          negativeFeedbackCount: 0,
        },
      ],
    },
  },
};

export const multipleReplies = {
  data: {
    GetArticle: {
      replyCount: 4,
      text: '老榮民九成存款全部捐給慈濟，如今窮了卻得不到慈濟醫院社工的幫忙，竟翻臉不認人',
      articleReplies: [
        {
          reply: {
            id: 'AV--O3nfyCdS-nWhujMD',
            type: 'RUMOR',
            text: '老榮民有住所，有月退俸，也不需要特別的醫療護理，因此較適合榮民之家養老，慈濟也一直關懷陪伴。護理之家有免費方案，入住資格和有無捐款沒有關係。\n慈善機構使用資源需符合規定，不能僅為少數人開特例。不是曾捐款就需要特別厚待，不是未曾捐款就可以不理。',
          },
          positiveFeedbackCount: 3,
          negativeFeedbackCount: 0,
        },
        {
          reply: {
            id: 'AV--LRZYyCdS-nWhujL9',
            type: 'NOT_ARTICLE',
            text: '整個基督教徒的打手，自己不努力去牧人、募款，利用攻擊別人，三分真七分假，更利用多帳號來造勢及造風向，害我也差點相信了，各位有智慧的看官，看過之後，笑一笑即可，別太認真去相信他的話啦。\n',
          },
          positiveFeedbackCount: 0,
          negativeFeedbackCount: 0,
        },
        {
          reply: {
            id: 'AV8jkRlByCdS-nWhuhiY',
            type: 'NOT_ARTICLE',
            text: '文章是假的 照片也是盜用我方家人的照片,嚴重影響我家人名譽,我方已對po文的人聯絡並打算提出告訴,請勿相信謠言.',
          },
          positiveFeedbackCount: 0,
          negativeFeedbackCount: 0,
        },
        {
          reply: {
            id: 'AV8fXikZyCdS-nWhuhfN',
            type: 'OPINIONATED',
            text: '慈濟自來有褒有貶，每人有各自的看法。',
          },
          positiveFeedbackCount: 0,
          negativeFeedbackCount: 0,
        },
      ],
    },
  },
};

export const elevenReplies = {
  data: {
    GetArticle: {
      text: 'eleven Replies article text',
      replyCount: 11,
      articleReplies: [
        {
          reply: {
            id: 'AV--O3nfyCdS-nWhujMD',
            type: 'RUMOR',
            text: '回應1',
          },
          positiveFeedbackCount: 0,
          negativeFeedbackCount: 0,
        },
        {
          reply: {
            id: 'AV--O3nfyCdS-nWhujMD',
            type: 'RUMOR',
            text: '回應2',
          },
          positiveFeedbackCount: 0,
          negativeFeedbackCount: 0,
        },
        {
          reply: {
            id: 'AV--O3nfyCdS-nWhujMD',
            type: 'NOT_RUMOR',
            text: '回應3',
          },
          positiveFeedbackCount: 0,
          negativeFeedbackCount: 0,
        },
        {
          reply: {
            id: 'AV--O3nfyCdS-nWhujMD',
            type: 'NOT_RUMOR',
            text: '回應4',
          },
          positiveFeedbackCount: 0,
          negativeFeedbackCount: 0,
        },
        {
          reply: {
            id: 'AV--O3nfyCdS-nWhujMD',
            type: 'RUMOR',
            text: '回應5',
          },
          positiveFeedbackCount: 0,
          negativeFeedbackCount: 0,
        },
        {
          reply: {
            id: 'AV--O3nfyCdS-nWhujMD',
            type: 'RUMOR',
            text: '回應6',
          },
          positiveFeedbackCount: 0,
          negativeFeedbackCount: 0,
        },
        {
          reply: {
            id: 'AV--LRZYyCdS-nWhujL9',
            type: 'NOT_ARTICLE',
            text: '回應7。\n',
          },
          positiveFeedbackCount: 0,
          negativeFeedbackCount: 0,
        },
        {
          reply: {
            id: 'AV8jkRlByCdS-nWhuhiY',
            type: 'NOT_ARTICLE',
            text: '回應8',
          },
          positiveFeedbackCount: 0,
          negativeFeedbackCount: 0,
        },
        {
          reply: {
            id: 'AV8fXikZyCdS-nWhuhfN',
            type: 'OPINIONATED',
            text: '回應9',
          },
          positiveFeedbackCount: 0,
          negativeFeedbackCount: 0,
        },
        {
          reply: {
            id: 'AV--O3nfyCdS-nWhujMD',
            type: 'RUMOR',
            text: '回應10',
          },
          positiveFeedbackCount: 0,
          negativeFeedbackCount: 0,
        },
        {
          reply: {
            id: 'AV--O3nfyCdS-nWhujMD',
            type: 'RUMOR',
            text: '回應11',
          },
          positiveFeedbackCount: 0,
          negativeFeedbackCount: 0,
        },
      ],
    },
  },
};

export const noReplies = {
  data: {
    GetArticle: {
      text: '老司機車裡總備一塊香皂，知道內情的新手默默也準備了一塊',
      articleType: 'TEXT',
      replyCount: 0,
      articleReplies: [],
    },
  },
};

export const shortArticle = {
  data: {
    GetArticle: {
      text: 'YouTube  ·  寻找健康人生',
      replyCount: 0,
      articleReplies: [],
    },
  },
};

export const oneImageArticle = {
  data: {
    GetArticle: {
      text: '',
      replyCount: 2,
      articleReplies: [
        {
          reply: {
            id: 'AVygFA0RyCdS-nWhuaXY',
            type: 'RUMOR',
            text: '查證這則謠言的過程中，約翰走鹿也發現其實網路早就有許多解答，就讓約翰走鹿來借花獻佛一下囉。多家媒體都有針對這則「吃生魚片會有海獸胃腺蟲附著胃壁」的謠言請教醫師，受訪的醫師們皆表示，這則謠言太誇大，海獸胃腺蟲沒那麼神通廣大。\n醫師們表示，台灣的衛生條件很進步，吃生魚片被寄生蟲寄生的案例很少，大部份都是生魚片不乾淨，感染沙門氏菌、大腸桿菌的案例，此外，寄生蟲也不是什麼難治之症，服用一些抗寄生蟲藥就可以殺死胃裡的這些蟲子，除非這些蟲子跑到身體其他部位那才比較難處理，但那也是要大量地吃生魚片（一週兩次以上），中獎機率才會比較高。',
          },
          positiveFeedbackCount: 14,
          negativeFeedbackCount: 2,
        },
        {
          reply: {
            id: 'AVy6LkWIyCdS-nWhuaqu',
            type: 'RUMOR',
            text: '過去台灣只有一個海獸胃腺蟲鑽入胃壁的病例，而且僅有一隻蟲。而經查詢，台北馬偕醫院也沒有名為林清風的醫師。',
          },
          positiveFeedbackCount: 24,
          negativeFeedbackCount: 0,
        },
      ],
    },
  },
};

export const createOrUpdateReplyRequest = {
  data: { CreateOrUpdateReplyRequest: { replyRequestCount: 1 } },
};
