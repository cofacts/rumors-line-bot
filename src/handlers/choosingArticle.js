import gql from '../gql';
import {
  createPostbackAction,
  createFeedbackWords,
  createTypeWords,
} from './utils';

const SITE_URL = process.env.SITE_URL || 'https://cofacts.g0v.tw/';

/**
 * 第2句 (template message)：按照時間排序「不在查證範圍」之外的回應，每則回應第一行是
 * 「⭕ 含有真實訊息」或「❌ 含有不實訊息」之類的 (含 emoticon)，然後是回應文字。如果
 * 還有空間，才放「不在查證範圍」的回應。最後一句的最後一格顯示「看其他回應」，連到網站。
 */
function reorderReplies(replyConnections) {
  const replies = [];
  const notArticleReplies = [];

  for (let replyConnection of replyConnections) {
    const { reply, feedbacks, id } = replyConnection;
    const item = {
      ...reply,
      feedbacks,
      replyConnectionId: id,
    };
    if (reply.type !== 'NOT_ARTICLE') {
      replies.push(item);
    } else {
      notArticleReplies.push(item);
    }
  }
  return replies.concat(notArticleReplies);
}

export default async function choosingArticle(params) {
  let { data, state, event, issuedAt, userId, replies, isSkipUser } = params;

  if (!data.foundArticleIds) {
    throw new Error('foundArticleIds not set in data');
  }

  data.selectedArticleId = data.foundArticleIds[event.input - 1];
  const { selectedArticleId } = data;

  if (+event.input === 0) {
    replies = [
      {
        type: 'template',
        altText: '請問要將文章送出到資料庫嗎？\n「是」請輸入「y」，「否」請輸入「n」或其他單一字母。',
        template: {
          type: 'buttons',
          text: '請問要將文章送出到資料庫嗎？',
          actions: [
            createPostbackAction('是', 'y', issuedAt),
            createPostbackAction('否', 'n', issuedAt),
          ],
        },
      },
    ];

    state = 'ASKING_ARTICLE_SUBMISSION';
  } else if (!selectedArticleId) {
    replies = [
      {
        type: 'text',
        text: `請輸入 1～${data.foundArticleIds.length} 的數字，來選擇文章。`,
      },
    ];

    state = 'CHOOSING_ARTICLE';
  } else {
    const { data: { GetArticle } } = await gql`
      query($id: String!) {
        GetArticle(id: $id) {
          replyCount
          replyConnections(status: NORMAL) {
            id
            reply {
              id
              versions(limit: 1) {
                type
                text
              }
            }
            feedbacks {
              comment
              score
            }
          }
        }
      }
    `({
      id: selectedArticleId,
    });

    const count = {};

    GetArticle.replyConnections.forEach(e => {
      const type = e.reply.versions[0].type;
      if (!count[type]) {
        count[type] = 1;
      } else {
        count[type]++;
      }
    });

    const articleReplies = reorderReplies(GetArticle.replyConnections);
    const summary =
      '這篇文章有：\n' +
      `${count.RUMOR || 0} 則回應認為其 ❌ 含有不實訊息\n` +
      `${count.NOT_RUMOR || 0} 則回應認為其 ⭕ 含有真實訊息\n` +
      `${count.OPINIONATED || 0} 則回應認為其 💬 含有個人意見\n` +
      `${count.NOT_ARTICLE || 0} 則回應認為其 ⚠️️ 不在查證範圍\n`;

    replies = [
      {
        type: 'text',
        text: summary,
      },
    ];

    if (articleReplies !== 0) {
      data.foundReplies = articleReplies.map(({ replyConnectionId, id }) => ({
        id,
        replyConnectionId,
      }));
      state = 'CHOOSING_REPLY';

      if (articleReplies === 1) {
        // choose for user
        event.input = 1;

        return {
          data,
          state: 'CHOOSING_REPLY',
          event,
          issuedAt,
          userId,
          replies,
          isSkipUser: true,
        };
      }

      replies.push({
        type: 'template',
        altText: articleReplies
          .slice(0, 10)
          .map(
            (
              { versions, feedbacks },
              idx
            ) => `閱讀請傳 ${idx + 1}> ${createFeedbackWords(feedbacks)} \n ${versions[0].text.slice(0, 20)}`
          )
          .join('\n\n'),
        template: {
          type: 'carousel',
          columns: articleReplies.map(({ versions, feedbacks }, idx) => ({
            text: createTypeWords(versions[0].type) +
              '\n' +
              createFeedbackWords(feedbacks) +
              '\n' +
              versions[0].text.slice(0, 80),
            actions: [createPostbackAction('閱讀此回應', idx + 1, issuedAt)],
          })),
        },
      });
      let endText;
      if (articleReplies.length > 10) {
        endText = '更多回應請到：';
      } else {
        endText = '詳情請見：';
      }
      replies.push({
        type: 'text',
        text: `${endText}${SITE_URL}/article/${selectedArticleId}`,
      });
    } else {
      // No one has replied to this yet.
      const { data: { CreateReplyRequest }, errors } = await gql`
        mutation($id: String!) {
          CreateReplyRequest(articleId: $id) {
            replyRequestCount
          }
        }
      `({ id: selectedArticleId }, { userId });

      replies = [
        {
          type: 'text',
          text: `目前還沒有人回應這篇文章唷。${errors ? '' : `已經將您的需求記錄下來了，共有 ${CreateReplyRequest.replyRequestCount} 人跟您一樣渴望看到針對這篇文章的回應。`}`,
        },
        {
          type: 'text',
          text: `若有最新回應，會寫在這個地方：${SITE_URL}/article/${selectedArticleId}`,
        },
      ];

      state = '__INIT__';
    }
  }

  return { data, state, event, issuedAt, userId, replies, isSkipUser };
}
