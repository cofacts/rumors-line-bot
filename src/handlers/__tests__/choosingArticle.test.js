import choosingArticle from '../choosingArticle';
import apiResult from '../__fixtures__/apiResult';
import * as gql from '../../gql';

describe('choosingArticle(params)', () => {
  it('should select article by articleId', done => {
    /* eslint-disable import/namespace */
    gql.default = () => () => {
      return new Promise(resolve => {
        resolve(apiResult.selectedArticleId);
      });
    };
    let params = {
      data: {
        searchedText: '《緊急通知》\n台北馬偕醫院傳來訊息：\n資深醫生（林清風）傳來：「請大家以後千萬不要再吃生魚片了！」\n因為最近已經發現- 好多病人因為吃了生魚片，胃壁附著《海獸胃腺蟲》，大小隻不一定，有的病人甚至胃壁上滿滿都是無法夾出來，驅蟲藥也很難根治，罹患機率每個國家的人都一樣。\n尤其；鮭魚的含蟲量最高、最可怕！\n請傳給朋友，讓他們有所警惕!',
        foundArticleIds: [
          'AVyyB61NyCdS-nWhuakC',
          'AV4sEcSHyCdS-nWhufEX',
          '5478658696099-rumor',
        ],
      },
      state: 'CHOOSING_ARTICLE',
      event: {
        type: 'message',
        input: '1',
        timestamp: 1505314293406,
        message: {
          type: 'text',
          id: '6691804381697',
          text: '1',
        },
      },
      issuedAt: 1505314295017,
      userId: 'Uc76d8ae9ccd1ada4f06c4e1515d46466',
      isSkipUser: false,
    };

    choosingArticle(params).then(
      result => {
        expect(result).toMatchSnapshot();
        done();
      },
      error => {
        console.log(error);
        expect(error).toBeUndefined();
      }
    );
  });
  it('should select article and have OPINIONATED and NOT_ARTICLE replies', done => {
    let params = {
      data: {
        searchedText: '老榮民九成存款全部捐給慈濟，如今窮了卻得不到慈濟醫院社工的幫忙，竟翻臉不認人',
        foundArticleIds: ['AV8d2-YtyCdS-nWhuhdi'],
      },
      state: 'CHOOSING_ARTICLE',
      event: {
        type: 'message',
        input: '1',
        timestamp: 1511633232479,
        message: { type: 'text', id: '7045918737413', text: '1' },
      },
      issuedAt: 1511633232970,
      userId: 'Uc76d8ae9ccd1ada4f06c4e1515d46466',
      replies: undefined,
      isSkipUser: false,
    };
    gql.default = () => () => {
      return new Promise(resolve => {
        resolve(apiResult.multipleReplies);
      });
    };
    choosingArticle(params).then(
      result => {
        expect(result).toMatchSnapshot();
        done();
      },
      error => {
        console.log(error);
        expect(error).toBeUndefined();
      }
    );
  });
});
