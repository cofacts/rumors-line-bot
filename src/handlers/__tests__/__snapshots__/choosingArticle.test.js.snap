exports[`choosingArticle(params) select article by articleId 1`] = `
Object {
  "data": Object {
    "foundArticleIds": Array [
      "AVyyB61NyCdS-nWhuakC",
      "AV4sEcSHyCdS-nWhufEX",
      "5478658696099-rumor",
    ],
    "foundReplies": Array [
      Object {
        "id": "AVygFA0RyCdS-nWhuaXY",
        "replyConnectionId": "AVyyB61NyCdS-nWhuakC__AVygFA0RyCdS-nWhuaXY",
      },
      Object {
        "id": "AVy6LkWIyCdS-nWhuaqu",
        "replyConnectionId": "AVyyB61NyCdS-nWhuakC__AVy6LkWIyCdS-nWhuaqu",
      },
    ],
    "searchedText": "《緊急通知》
台北馬偕醫院傳來訊息：
資深醫生（林清風）傳來：「請大家以後千萬不要再吃生魚片了！」
因為最近已經發現- 好多病人因為吃了生魚片，胃壁附著《海獸胃腺蟲》，大小隻不一定，有的病人甚至胃壁上滿滿都是無法夾出來，驅蟲藥也很難根治，罹患機率每個國家的人都一樣。
尤其；鮭魚的含蟲量最高、最可怕！
請傳給朋友，讓他們有所警惕!",
    "selectedArticleId": "AVyyB61NyCdS-nWhuakC",
  },
  "event": Object {
    "input": "1",
    "message": Object {
      "id": "6691804381697",
      "text": "1",
      "type": "text",
    },
    "timestamp": 1505314293406,
    "type": "message",
  },
  "isSkipUser": false,
  "issuedAt": 1505314295017,
  "replies": Array [
    Object {
      "text": "這篇文章有：
2 則回應認為其 ❌ 含有不實訊息
0 則回應認為其 ⭕ 含有真實訊息
0 則回應認為其 💬 含有個人意見
0 則回應認為其 ⚠️️ 不在查證範圍
",
      "type": "text",
    },
    Object {
      "altText": "閱讀請傳 1> [有 14 人覺得此回應有幫助
有 2 人覺得此回應沒幫助] 
 查證這則謠言的過程中，約翰走鹿也發現其實

閱讀請傳 2> [有 24 人覺得此回應有幫助] 
 過去台灣只有一個海獸胃腺蟲鑽入胃壁的病例",
      "template": Object {
        "columns": Array [
          Object {
            "actions": Array [
              Object {
                "data": "{\"input\":1,\"issuedAt\":1505314295017}",
                "label": "閱讀此回應",
                "type": "postback",
              },
            ],
            "text": "❌ 含有不實訊息
[有 14 人覺得此回應有幫助
有 2 人覺得此回應沒幫助]
查證這則謠言的過程中，約翰走鹿也發現其實網路早就有許多解答，就讓約翰走鹿來借花獻佛一下囉。多家媒體都有針對這則「吃生魚片會有海獸胃腺蟲附著胃壁」的謠言請教醫師，",
          },
          Object {
            "actions": Array [
              Object {
                "data": "{\"input\":2,\"issuedAt\":1505314295017}",
                "label": "閱讀此回應",
                "type": "postback",
              },
            ],
            "text": "❌ 含有不實訊息
[有 24 人覺得此回應有幫助]
過去台灣只有一個海獸胃腺蟲鑽入胃壁的病例，而且僅有一隻蟲。而經查詢，台北馬偕醫院也沒有名為林清風的醫師。",
          },
        ],
        "type": "carousel",
      },
      "type": "template",
    },
    Object {
      "text": "詳情請見：https://cofacts.g0v.tw//article/AVyyB61NyCdS-nWhuakC",
      "type": "text",
    },
  ],
  "state": "CHOOSING_REPLY",
  "userId": "Uc76d8ae9ccd1ada4f06c4e1515d46466",
}
`;
